scipy/optimize/_linprog_ip.py

```diff
@@ -0,0 +1,1940 @@
+             1 : Iteration limit reached
+             2 : Problem appears to be infeasible
+             3 : Problem appears to be unbounded
+             4 : Serious numerical difficulties encountered
+    message : str
+        A string descriptor of the exit status of the optimization.
+
+    """
+    # note that all the inputs are the ORIGINAL, unmodified versions
+    # no rows, columns have been removed
+    # the only exception is bounds; it has been modified
+    # we need these modified values to undo the variable substitutions
+    # in retrospect, perhaps this could have been simplified if the "undo"
+    # variable also contained information for undoing variable substitutions
+
+    n_x = len(c)
+
+    # we don't have to undo variable substitutions for fixed variables that
+    # were removed from the problem
+    no_adjust = set()
+
+    # if there were variables removed from the problem, add them back into the
+    # solution vector
+    if len(undo) > 0:
+        no_adjust = set(undo[0])
+        x = x.tolist()
+        for i, val in zip(undo[0], undo[1]):
+            x.insert(i, val)
+        x = np.array(x)
+
+    # now undo variable substitutions
+    # if "complete", problem was solved in presolve; don't do anything here
+    if not complete and bounds is not None:  # bounds are never none, probably
+        n_unbounded = 0
+        for i, b in enumerate(bounds):
+            if i in no_adjust:
+                continue
+            lb, ub = b
+            if lb is None and ub is None:
+                n_unbounded += 1
+                x[i] = x[i] - x[n_x + n_unbounded - 1]
+            else:
+                if lb is None:
+                    x[i] = ub - x[i]
+                else:
+                    x[i] += lb
+
+    n_x = len(c)
+    x = x[:n_x]  # all the rest of the variables were artificial
+    fun = x.dot(c)
+    slack = b_ub - np.dot(A_ub, x)  # report slack for ORIGINAL UB constraints
+    # report residuals of ORIGINAL EQ constraints
+    con = b_eq - np.dot(A_eq, x)
+
+    if status == 0 and (np.isnan(x).any() or np.isnan(fun) or
+                        np.isnan(slack).any() or np.isnan(con).any()):
+        status = 4
+        message = "Numerical difficulties were encountered but no errors " \
+                  "were raised. This is known to occur if the 'presolve' " \
+                  "option is False, 'sparse' is True, and A_eq includes " \
+                  "redundant rows. If you encounter this under different " \
+                  "circumstances, please submit a bug report. Otherwise, " \
+                  "remove linearly dependent equations from your equality " \
+                  "constraints or enable presolve."
+
+    return x, fun, slack, con, status, message
+
+
+def _get_solver(sparse=False, lstsq=False, sym_pos=False, cholesky=False):
+    """
+    Given solver options, return a handle to the appropriate linear system
+    solver.
+
+    Parameters
+    ----------
+    sparse : boolean
```

Comment:
docstring format: the type descriptor should be ``bool`` rather than ``boolean``. In descriptions ``boolean`` does make sense of course.